### PR TITLE
expression: fix collate bug in constant propagation 

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve? 
We can't propagate constant if two column's collations are different. Here is a example:
```
create table t (a char(10) collate utf8mb4_bin, b char(10) collate utf8mb4_general_ci);
insert into t values ('a', 'A');
select * from t t1, t t2 where t1.a=t2.b and t2.b='a';
```
We can't rewrite `t1.a=t2.b and t2.b='a'` to `t1.a='a' and t2.b='a'`. If so, the query will return one line instead of empty.

### What is changed and how it works?
When try to propagate constant, check if the collation are consistent.

### Check List 

Tests 

 - Integration test

Code changes

Side effects

Related changes

Release note

### CODE REVIEW

File: types/field_type.go [hunk 1]
```diff
@@ -35,11 +35,15 @@ type FieldType = ast.FieldType
 // NewFieldType returns a FieldType,
 // with a type and other information about field type.
 func NewFieldType(tp byte) *FieldType {
-	return &FieldType{
+	ft := &FieldType{
 		Tp:      tp,
 		Flen:    UnspecifiedLength,
 		Decimal: UnspecifiedLength,
 	}
+	if tp != mysql.TypeVarchar && tp != mysql.TypeVarString && tp != mysql.TypeString {
```
Comment (bb7133) [inline types/field_type.go]: Please add a TODO here for the default charset/collation of string-related types.

Comment (wjhuang2016) [reply types/field_type.go]: Already set the default collation for string-related types.

File: expression/constant_propagation.go [hunk 2]
```diff
@@ -197,7 +235,7 @@ func (s *propConstSolver) propagateColumnEQ() {
 		if fun, ok := s.conditions[i].(*ScalarFunction); ok && fun.FuncName.L == ast.EQ {
 			lCol, lOk := fun.GetArgs()[0].(*Column)
 			rCol, rOk := fun.GetArgs()[1].(*Column)
-			if lOk && rOk {
+			if lOk && rOk && lCol.GetType().Collate == rCol.GetType().Collate {
```
Comment (qw4990) [inline expression/constant_propagation.go]: Using `CompatibleCollate` is better?

File: expression/constant_propagation.go [hunk 3]
```diff
@@ -166,6 +190,20 @@ func (s *propConstSolver) propagateConstantEQ() {
 		}
 		for i, cond := range s.conditions {
 			if !visited[i] {
+				// Make sure the collations of the argument are equal.
+				if sf, ok := cond.(*ScalarFunction); ok && len(sf.GetArgs()) > 1 {
+					baseCollation := sf.GetArgs()[0].GetType().Collate
+					skip := false
+					for _, arg := range sf.GetArgs() {
+						if arg.GetType().Collate != baseCollation {
```
Comment (qw4990) [inline expression/constant_propagation.go]: ditto, it's better to use `CompatibleCollate` here?

**Pull request merged**